plutus-core/plutus-core/src/PlutusCore/MkPlc.hs

```diff
@@ -339,5 +340,11 @@ mkFreshTermLet aT a = do
     pure (var mempty genName, termLet mempty (Def (VarDecl mempty genName aT) a))
 
 -- | 'apply' the head of the application to the arguments iteratively.
-headSpineToTerm :: TermLike term tyname name uni fun => HeadSpine (term ()) -> term ()
-headSpineToTerm = foldl1 (apply ())
+headSpineToTerm :: TermLike term tyname name uni fun => ann -> MonoHeadSpine (term ann) -> term ann
+headSpineToTerm _ (HeadOnly t)       = t
+headSpineToTerm ann (HeadSpine t ts) = apply ann t $ foldl1 (apply ann) ts
```

Comment:
This doesn't look right at all. Does it not fail any tests?

Just use `foldl`.

Reply:
brainfart!

Reply:
I need to know if it failed any tests! 'Cause if it didn't, that's a problem on its own.